### PR TITLE
feat(misc): Keyboard handling helper

### PR DESCRIPTION
**Pull request by kmcfaul**
Closes: #5864

I've got 2 helper options I've been playing with and am wondering what people think.

Option 1: helper function. This is a function we'd be able to inject into components' local key handlers that covers basic arrow key navigation (and I can pull in basic Space/Enter key as well). Various optional props in the function allow of customization of what the handler covers.

Option 2: helper component. This is a renderless component I was experimenting with that would handle the set up/tear down of event listeners as well as the basic navigation, with additional key handling (and other customizations) that can be passed as props. It feels a bit odd to me plopping a random component inside the parent (shown in Menu), but since it's renderless it doesn't show up in the DOM or our testing snapshots.

For now, DualListSelector and Menu have been updated to use the new options (1 in DualListSelector, 2 in Menu). The commented code in Menu shows Option 1 before I was trying out the new component. I've half-converted Select, but that is set up a fair bit differently than other keyboard handlers (i.e. not using a global handler, but instead each item having its own key handler) so there's a bit more work to be done and I've left it out the PR. 

PR also includes 2 minor fixes regarding Menu keyboard navigation:
1. Drilldown menu transition triggered with keyboard now properly shifts focus to sub-menu/parent menu.
2. Menu footer now correctly has a tabIndex, which was being erroneously removed (selector needed an update).

### CODE REVIEW

**Review by mcarrano — COMMENTED** [review]
@kmcfaul this looks good to me except that I was able to hang the page in Firefox when I was playing with the drilldown menu. Not sure that I can reproduce. 

As to your questions, not sure I'm the best one to advise on the best coding approach. Would defer to @tlabaj @nicolethoen and @jessiehuff for more feedback.

**Comment by kmcfaul** [comment]
@mcoker Did the transition properties get updated for drilldown menus? I previously only listened for a `transitionend` event that had an event type of `visibility` when changing focus for keyboard, but now it seems no `visibility` updates on transition.

**Comment by mcoker** [comment]
@kmcfaul yep, `visibility` was removed from `--pf-c-menu--m-drilldown--c-menu--Transition` in https://github.com/patternfly/patternfly/pull/4090/files#diff-e0945357fb0f2287914773cdb32f4c0f9b7b22e4c3e1cbd96fdcc45d254763c1L119-R115

**Comment by kmcfaul** [comment]
Thanks for confirming! I added the fix in this PR, just need to figure out that what I changed didn't introduce the inconsistent lag for the drilldown demo haha

**Comment by tlabaj** [comment]
This is great Katie!  I like the improvements you made since we had the code walkthrough meeting.

File: packages/react-core/src/helpers/KeyboardHandler.tsx
Comment (tlabaj) [inline packages/react-core/src/helpers/KeyboardHandler.tsx:32]: I might just expand on the description her and be clear about when you would use this function over the class.

**Review by nicolethoen — COMMENTED** [review]
Im seeing an issue when I try to tab out of a flyout menu - i think tabbing out of a flyout menu should close the flyout menu and return the tab index to the top level item?

I also am seeing the same hanging issue with the drilldown menu
I triggered it by hitting enter to try and open the drilldown menu? Maybe it is trying to select and drill down on enter and it's hanging because it doesn't know what to select?

File: packages/react-core/src/components/DualListSelector/DualListSelector.tsx
Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:612]: The component has some default values. The function can as well, though you really only save lines of code if you don't need any of the other props of the function (as you would have to pass in `undefined` for every skipped property that you don't override the default).

The `Tab` key isn't handled by the Menu component at the moment, instead being defined in the composable menu examples (as they can open/close). We don't yet have an example with flyout but I think it should be possible to have this behavior when we go to add it, if its not already included as I think Menu is re-mounted when it opens/closes so tab should be reset already. I'm not sure if there's a benefit in locally defining a `Tab` interaction as that would conflict with typical Tab behavior in composable menus?

Where are you seeing the issue with the drilldown demo? If it's in the composable menu demo for drilldown, the `Enter` key currently only opens/closes the menu. You have to `ArrowDown` to shift focusing on the menu, but that's something we could update. In the menu example, Enter appears to transition and focus correctly for me.

Comment (kmcfaul) [reply packages/react-core/src/components/DualListSelector/DualListSelector.tsx:612]: Added in a default for `getFocusableElement` and `isActiveElement` in the function handler while I was there updating the description. It doesn't hurt anything, though the default of `isActiveElement` maybe makes an extra document.activeElement call opposed to passing in a function that references a variable with that data stored.

**Comment by kmcfaul** [comment]
Added a small refactor of TreeView (non-checkbox) for vertical arrow keys. Checkbox arrow keys, and horizontal arrow keys are very custom to TreeView so are not refactored.

File: packages/react-core/src/helpers/KeyboardHandler.tsx
Comment (tlabaj) [inline packages/react-core/src/helpers/KeyboardHandler.tsx:32]: perfect!

**Comment by nicolethoen** [comment]
I think I figured out how to make the drilldown menu break - when I get to a menu item that can drill down, if I accidentally use the left/right arrow key to try and drill down instead of hitting enter, then it hangs. There is no console out put.

This video shows the hanging though. It just gets stuck and I can't do anything. Can't move focus, can't open console or refresh page, etc. Only option is to close window.

https://user-images.githubusercontent.com/26305082/125346150-64d76700-e327-11eb-998d-8af73b1ec0b2.mp4

**Comment by kmcfaul** [comment]
@nicolethoen Thank you so much for the info, I was able to reproduce the issue now! There's an infinite loop with left/right arrows for this menu (because the sibling of the drill in button is a full menu). Working on debugging and resolving the problem!

**Review by nicolethoen — APPROVED** [review]
LGTM :)

**Review by tlabaj — APPROVED** [review]
LGTM.  Lets get @jessiehuff review before we merge this one

**Review by jessiehuff — APPROVED** [review]
This is looking great! I'm also not noticing the weird page freezing on drilldown anymore. I noticed that the dual list expandable options with checkboxes wasn't completely keyboard accessible (I couldn't expand any of the options with keyboard), but when I tested with org I noticed that that's a pre-existing issue so we could open a separate issue for that.

**Pull request merged**
